File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int pre = 1;
        int suf = 1;
        int maxx = INT_MIN;
        int n = nums.size();

        for(int i = 0; i < n; i++){
            if(pre == 0) pre = 1;
            if(suf == 0) suf = 1;

            pre = pre * nums[i];
            suf = suf * nums[n-i-1];

            maxx = max(maxx, max(pre, suf));  
        }

        return maxx;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        pre = 1
        suf = 1
        max_prod = -1e9
        n = len(nums)

        for i in range(n):
            if pre == 0:
                pre = 1
            if suf == 0:
                suf = 1
            pre*=nums[i]
            suf*=nums[n-i-1]

            max_prod = max(max_prod, max(pre, suf))

        return max_prod
